File: src/main/java/com/zuehlke/depanalyzer/jdeps/JDepsException.java
package com.zuehlke.depanalyzer.jdeps;

import java.io.File;
import java.io.IOException;

public class JDepsException extends RuntimeException {
    private JDepsException(String message, Throwable e) {
        super(message, e);
    }

    static JDepsException runtimeError(String writerOutput) {
        return new JDepsException("Error while running jDeps: " + writerOutput, null);
    }

    static JDepsException targetFileDoesNotExist(File targetFile) {
        return new JDepsException("File " + targetFile.getAbsolutePath() + " does not exist", null);
    }

    static JDepsException targetFileNotReadable(File targetFile) {
        return new JDepsException("File " + targetFile.getAbsolutePath() + " is not readable", null);
    }

    static JDepsException wrap(Exception e) {
        if (e instanceof JDepsException) {
            return (JDepsException) e;
        }
        return new JDepsException("Error while running jDeps: " + e.getMessage(), e);
    }

    static JDepsException noFatJar(String fileName) {
        return new JDepsException(fileName + " is not a fat jar", null);
    }

    static JDepsException couldNotReadFatJar(IOException e) {
        return new JDepsException("Could not read fat jar", e);
    }

    public static JDepsException cannotUnzip(String message, Throwable cause) {
        return new JDepsException("Could not unzip fat jar: " + message, cause);
    }

    public static JDepsException cannotUnzip(Throwable cause) {
        return cannotUnzip(cause.getMessage(), cause);
    }
}

File: java-dependency-analyzer/src/main/kotlin/site/gutschi/dependency/jdeps/JDepsException.kt
package site.gutschi.dependency.jdeps

import java.io.File

class JDepsException private constructor(message: String, e: Throwable?) : RuntimeException(message, e) {
    companion object {
        fun targetFileDoesNotExist(targetFile: File): JDepsException {
            return JDepsException("File ${targetFile.absolutePath} does not exist", null)
        }

        fun targetFileNotReadable(targetFile: File): JDepsException {
            return JDepsException("File ${targetFile.absolutePath} is not readable", null)
        }

        fun wrap(e: Exception): JDepsException {
            return if (e is JDepsException) e else JDepsException("Error while running jDeps: ${e.message}", e)
        }

        fun cannotUnzip(cause: Throwable): JDepsException {
            return JDepsException("Could not unzip fat jar: ${cause.message}", cause)
        }

        fun matcherAfterFile(): Throwable {
            return JDepsException("Cannot change fatJarMatcher after files have been added", null)
        }

        fun invalidOutputLine(line: String): Throwable {
            return JDepsException("Invalid output line $line cannot be parsed", null)
        }

        fun nonZeroReturn(result: Int, output: String): Throwable {
            return JDepsException("JDeps returned $result: $output", null)

        }

        fun jdepsNotFound(): Throwable {
            return JDepsException("JDeps not a registered tool, please use a regular JVM", null)
        }

    }
}